### DISCUSSION TITLE
react-markdown : where can i find the react-markdown type source code?

### DISCUSSION THREAD

**Discussion by moonkorea00** [remarkjs/remark#1260] [Q&A] [answered]
Hey team,

regarding react-markdown, i'm trying to take a look into how the types are configured in the [react-markdown](https://github.com/remarkjs/react-markdown) repository.

for instance, i can find ast-to-react.d.ts under the node_modules/react-markdown directory in my project where i installed the library but where can i find it in the react-markdown repository?

```ts
// where can i find the type for SpecialComponents in the source code?
export type SpecialComponents = {
  code: CodeComponent | ReactMarkdownNames
  h1: HeadingComponent | ReactMarkdownNames
  h2: HeadingComponent | ReactMarkdownNames
  h3: HeadingComponent | ReactMarkdownNames
  h4: HeadingComponent | ReactMarkdownNames
  h5: HeadingComponent | ReactMarkdownNames
  h6: HeadingComponent | ReactMarkdownNames
  li: LiComponent | ReactMarkdownNames
  ol: OrderedListComponent | ReactMarkdownNames
  td: TableCellComponent | ReactMarkdownNames
  th: TableCellComponent | ReactMarkdownNames
  tr: TableRowComponent | ReactMarkdownNames
  ul: UnorderedListComponent | ReactMarkdownNames
}
```

thanks.

**Comment by ChristianMurphy** [comment, marked as answer]
Welcome @moonkorea00!

All of the types are in the source JavaScript files, inside JSDoc comments https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html

**Reply by moonkorea00** [reply]
thanks @ChristianMurphy.

i was just wondering because i created a custom type similar to `HeadingProps` for the heading elements where i gave the `level` property's type as `1 | 2 | 3 | 4 | 5 | 6`, however, `Options.components` required the level property as a number type.

or is this non-trivial that i just follow react-markdown's `HeadingProps` :
```ts
export type HeadingProps = ComponentPropsWithoutRef<'h1'> &
  ReactMarkdownProps & {
    level: number
  }
```

**Reply by ChristianMurphy** [reply]
@moonkorea00 that looks like an older version of `react-markdown` consider upgrading to the latest version (currently version 9) for better types and a cleaner API.
